### ISSUE TITLE
Expand capabilities for playing MP3, OGG or WMA file types to include looping and layering.

### ISSUE THREAD

**Issue by AugsEU** [MonoGame/MonoGame#8589]
### Intent

If I have a sound file in one of MP3, OGG or WMA  formats. I should be able to load it into my data and then loop and layer them. These formats are usually used for background music.

**Looping:** The ability to play a song and have it seamlessly jump to an earlier point in the song. Seamlessly meaning that there is no gap in audio. The most common example is going back to the start of a song when we reach the end. Ideally, there should also be the possibility of looping certain segments within a song. For example I might have an intro section that plays only once, then have the loop in the middle of my song.

**Layering:** The ability to play multiple songs at once. For example, I could have an mp3 which is just a drum loop, and a separate file which is just guitar. I should then be able to combine them at runtime in MonoGame by playing both at the same time. Another common usecase would be fading between two songs, as one fades out, another fades in. But for a breif time they are both playing at the same time.

### Motivation

If I have a long audio file in MonoGame, such as for background music, there are no good options for the developer.

**Option 1:** Use the MediaPlayer. While this is capable of playing the mentioned formats, it is not capable of looping seamlessly. Using the IsRepeating property does loop the song, but not seamlessly. There is a notceable gap inbetween loops. The MediaPlayer is also not capable of layering; it can only play 1 song at a time.

**Option 2:** Load your song as a SoundEffect. SoundEffects are capable of seamlessly looping, and are capable of layering. I can play multiple sound effects at the same time. However, this option is not good for long audio files as SoundEffect only supports uncompressed .wav files. This is not good because .wav files are very large. For context, my latest game which has a 12 song soundtrack would have had 0.5GB of sound files. Some games might have up to 50 songs, putting this number well above 1GB. If we wanted to do dynamic mixing and separate out the layers into individual tracks, that might quadruple the file size again.

**Looping** is a necessary feature for any game. Most games will have the player able to be in one level/zone for an indefinite amount of time, thus we need a way to play music continuously.

**Layering** is a nice-to-have; it lets developers be more creative with their sound design, possibly adding dynamic elements. Most modern games use this feature, and developers will expect it from any framework.

I have made a simple test project: https://github.com/AugsEU/MonoGameSongProblem . It contains the same short song as an mp3 and as a wav file. You can see that the MediaPlayer is not able to loop the song seamlessly. You can turn on playSongAsSFX to play it as a sound effect and observe what it sounds like when looped seamlessly.

**Comment by AugsEU** [comment]
Note: It seems to me that the best way to do this would be to add the possibility of loading a SoundEffect from a compressed audio format such as mp3.

**Comment by dellis1972** [comment]
@tomspilman @mrhelmut thoughts on this one?

**Comment by squarebananas** [comment]
Great writeup. Here are a few of my thoughts for now:

> Option 1: Use the MediaPlayer. While this is capable of playing the mentioned formats, it is **_not capable of looping seamlessly_**.

Tom's native plan [here](https://github.com/MonoGame/MonoGame/pull/8309#discussion_r1740343020) is to remove the current device media player system entirely. I did suggest a `PlatformBehavior` setting for playback via device media player or `SoundEffect`s but the response was _"I think this is would be a waste of MG development resources. I've seen maybe 2 or 3 MG/XNA games in the last 10 years that wanted to integrate with the device media player"_. I kind of agree (that means no Android/IOS support of game radio stations via URI though, someone see if [TruckerFM](https://truckers.fm/) works).

>Option 2: Load your song as a SoundEffect. SoundEffects are capable of seamlessly looping, and are capable of layering. I can play multiple sound effects at the same time. However, this option is not good for long audio files as SoundEffect only supports uncompressed .wav files.

Technically compression should be supported via [ADPCM](https://learn.microsoft.com/en-us/windows/win32/xaudio2/adpcm-overview) as well. MonoGame [issue here](https://github.com/MonoGame/MonoGame/issues/6701#issuecomment-2254123931) and [PR here](https://github.com/MonoGame/MonoGame.Tool.FFmpeg/pull/6).

### General Audio Suggestions
- Create a table with columns for each available audio system (`SoundEffect`, `Song`, XACT's `Cue`) and rows for all functionality features (panning, reverb, visualizer, etc) to discover which system has the most functionality already.
- Add all missing functionality to one system (probably `SoundEffect`).
- Make the other systems wrappers around this all-in-one base class.
- For native work, only implement the all-in-one base audio system and make the other audio systems C# wrappers around this.
- Over time bring missing features to the more limited audio systems.

**Comment by AugsEU** [comment]
Using ADPCM does alleviate the problem somewhat. I have tested with a 3m32s song.

WAV: 38.8MB

XNB-Best: 38.8MB
XNB-Medium: 7.34MB
XNB-Low: 4.89MB

OGG: 2.86MB

So it is about 90% more compressed than a raw wav, but still twice the size of OGG, and likely lower quality(but this is subjective).

However, it is notable that this is still not a viable option because the compressed ADPCM sound effects can't loop. See: https://github.com/MonoGame/MonoGame/issues/8433

Also SoundEffect doesn't support looping sections within a sound, it only supports looping the sound as a whole.

**Comment by squarebananas** [comment]
> However, it is notable that this is still not a viable option because the compressed ADPCM sound effects can't loop. See: #8433

Ok, I wasn't fully aware of that. I need to read the ADPCM overview I linked more, but I did notice a couple of interesting points:

"Larger blocks allow better compression, which results in smaller file sizes, but at the expense of sound quality and resolution for aligning loop points."
(https://learn.microsoft.com/en-us/windows/win32/xaudio2/adpcm-overview#samplesperblock)

"The value of the SamplesPerBlock parameter affects the resolution with which you can align wave data and loop points."
(https://learn.microsoft.com/en-us/windows/win32/xaudio2/adpcm-overview#restrictions)

> Also SoundEffect doesn't support looping sections within a sound, it only supports looping the sound as a whole.

I would need to check another time but I think XACT supports those loop points. If it does they could be brought across at some stage maybe.

Also something I realised just, WAV doesn't need to be decompressed whereas ADPCM has a small decompression cost and OGG a much higher cost. With layering tracks with OGG compared to ADPCM, I wonder if that could be a performance issue, maybe just for lower spec devices like mobile.

**Comment by AugsEU** [comment]
There would be multiple ways of doing this:

1) Decompress into memory on load. So when you load the ogg file using the content manager, it will decompress the whole PCM in one buffer. This would allow for quick and easy playback. This means we still get the small size on disk. The obvious drawback is that this increases memory usage.

2) Decompress as we stream the file. So in this case we would decompress on-demand was we queue in the OpenAL buffers. This would be more complicated to implement and have some CPU cost. However, I have worked on some games where this was a net performance gain. The reason being that we have much less memory throughput, and that out-weighs the extra work for the CPU. You must consider that loading from the RAM isn't free, so loading compressed data is faster in that regard.

**Comment by squarebananas** [comment]
Any chance you could try unzipping this file into the `MonoGameSongProblem-master\bin\Debug\net6.0\Content` folder and this load code:
```
System.IO.Stream SongAsSFXStream = System.IO.File.Open(System.IO.Path.Combine("Content", "SongAsSFX.wav"), System.IO.FileMode.Open);
mSongAsSoundEffect = SoundEffect.FromStream(SongAsSFXStream);
```
[SongAsSFX.zip](https://github.com/user-attachments/files/17894047/SongAsSFX.zip)

**Comment by AugsEU** [comment]
The file @squarebananas provided is a compressed .wav version of the sample song. The original uncompressed wav was 2.92MB but this is 308kb. It does loop correctly, so this is a way to have compressed audio looping in MonoGame.

However it is worth noting that the quality is significantly degraded, to the point this wouldn't be acceptable for most developers. Compared to the mp3, it has a noticeable hiss. The mp3 is also smaller at only 250kb
